Source/WTF/wtf/text/AdaptiveStringSearcher.h

```diff
@@ -65,51 +65,30 @@ class AdaptiveStringSearcherBase {
     static constexpr bool exceedsOneByte(LChar) { return false; }
     static constexpr bool exceedsOneByte(UChar c) { return c > 0xff; }
 
-    template <typename T, typename U>
-    static inline T alignDown(T value, U alignment)
-    {
-        return reinterpret_cast<T>((reinterpret_cast<uintptr_t>(value) & ~(alignment - 1)));
-    }
-
-    static constexpr uint8_t getHighestValueByte(LChar character) { return character; }
-
-    static constexpr uint8_t getHighestValueByte(UChar character)
-    {
-        return std::max<unsigned>(static_cast<uint8_t>(character & 0xFF), static_cast<uint8_t>(character >> 8));
-    }
-
     template <typename PatternChar, typename SubjectChar>
     static inline int findFirstCharacter(std::span<const PatternChar> pattern, std::span<const SubjectChar> subject, int index)
     {
         const auto* subjectPtr = subject.data();
-        const PatternChar patternFirstChar = pattern[0];
-        const int maxN = (subject.size() - pattern.size() + 1);
+        PatternChar patternFirstChar = pattern[0];
 
-        if (sizeof(SubjectChar) == 2 && !patternFirstChar) {
-            // Special-case looking for the 0 char in other than one-byte strings.
-            // memchr mostly fails in this case due to every other byte being 0 in text
-            // that is mostly ascii characters.
-            for (int i = index; i < maxN; ++i) {
-                if (!subjectPtr[i])
-                    return i;
-            }
-            return -1;
-        }
-        const uint8_t searchByte = getHighestValueByte(patternFirstChar);
-        const SubjectChar searchChar = static_cast<SubjectChar>(patternFirstChar);
-        int pos = index;
-        do {
-            ASSERT(maxN - pos >= 0);
-            const SubjectChar* charPos = reinterpret_cast<const SubjectChar*>(memchr(subjectPtr + pos, searchByte, (maxN - pos) * sizeof(SubjectChar)));
-            if (charPos == nullptr)
+        if constexpr (sizeof(PatternChar) == 2 && sizeof(SubjectChar) == 1) {
+            if (!isLatin1(patternFirstChar))
```

Comment:
Why could we handle this case before and not now?

Reply:
Because the previous one was using memchr and a hack for both 16 / 8 bit cases. Now the logic is completely different.